### ISSUE TITLE
How to support Lazy Preparation in AdsMediaSource?

### ISSUE THREAD

**Issue by Romantic-LiXuefeng** [google/ExoPlayer#5141]
With a live HLS stream,  If I configure multiple pre-roll ads, pre-load the content,when the pre-roll ads have played, there exists an exception that the ts does not exist in server or `BehindLiveWindowException`.

**Comment by Romantic-LiXuefeng** [comment]
@ojw28 Did you already had the time to look into this issue? Is there anything I can do to support you?

**Comment by andrewlewis** [comment]
What ExoPlayer version are you using? This may have been fixed by 81a9293072faabdd95f241f84b29e32217982ac9. What do you mean by pre-load the content?

Please provide complete information as requested in the issue template. The issue template can be found [here](https://github.com/google/ExoPlayer/blob/release-v2/ISSUE_TEMPLATE).

**Comment by Romantic-LiXuefeng** [comment]
> This may have been fixed by 81a9293

I think fixed by 81a9293. Project forward the loading position may avoids `BehindLiveWindowException` . but there may exists the error ,the request ts segment has not exist in server.
For example, when a ts segment is in the process of being downloaded , and the network environment is poor at this time. Maybe the request ts segment has been deleted in server. The player will play failed. I think the player need to resolve this issue. Just like the `BehindLiveWindowException`, project forward the loading position.   But I did't find the associate codes. 
Now I custom the new `LoadErrorHandlingPolicy`, when the retry count >= the default retry count, we advance the loading position . 
### HlsSampleStreamWrapper.onLoadError()
```
      if (loadErrorHandlingPolicy instanceof NewTvLoadErrorHandlingPolicy) {
            if (errorCount == loadErrorHandlingPolicy.getMinimumLoadableRetryCount(loadable.type)) {
                if (isMediaChunk && bytesLoaded == 0) {
                    // First remove the load error chunk from chunk list
                    HlsMediaChunk removed = mediaChunks.remove(mediaChunks.size() - 1);
                    Assertions.checkState(removed == loadable);
                    if (mediaChunks.isEmpty()) {
                        pendingResetPositionUs = lastSeekPositionUs;
                    }
                }
                pendingResetPositionUs = lastSeekPositionUs;
                loadErrorAction = Loader.DONT_RETRY;
                // Second continue loading the new chunk
                if (!prepared) {
                    continueLoading(lastSeekPositionUs);
                } else {
                    callback.onContinueLoadingRequested(this);
                }
            } else {
                loadErrorAction = Loader.createRetryAction(/* resetErrorCount= */ false, retryDelayMs);
            }
            if (retryDelayMs == C.TIME_UNSET) {
                loadErrorAction = Loader.DONT_RETRY_FATAL;
            }
    }else {
            loadErrorAction =
                      retryDelayMs != C.TIME_UNSET
                          ? Loader.createRetryAction(/* resetErrorCount= */ false, retryDelayMs)
                          : Loader.DONT_RETRY_FATAL;
      }           
```
I hope you give me some advice.

**Comment by Romantic-LiXuefeng** [comment]
> This may have been fixed by 81a9293

Using the buffered duration to project forward the loading position. There may still cause the `BehindLiveWindowException`. 
Pre-load means that when the ad is playing and the `LoadControl` allows to load the next source.

Pre-load the living stream is not safe.  There is always the possibility of  an `BehindLiveWindowException` or a error that the request ts has been deleted in server.
So I think when the ad is playing, The player does not have to pre-download the living stream.  

Now I used two players , one to play the pre-ads, when the pre-ads have played ended, use another to play the living stream  .

**Comment by andrewlewis** [comment]
Imagine we are starting to load the content while still playing the preroll. We project forwards the live edge of the content when choosing the loading position. If the segment at the projected load position is not available yet we don't try to load anything, and retry later. Once it is available we start buffering at this projected load position, which means that (assuming playback can continue linearly) when we've finished playing the ad we are at the live edge.

> So I think when the ad is playing, The player does not have to pre-download the living stream.

I don't follow how you arrived at this conclusion so please could you rephrase/clarify? There's also general advice on avoiding `BehindLiveWindowException`s covered in various other issues on this tracker so you may want to look at those.

Using separate players for ads and content is not a good solution as it can cause spiking resource usage among other issues.

**Comment by andrewlewis** [comment]
Closing due to inactivity.
